Three-level model caching

smallapes opened the pull request:
## Description

This PR implements an elastic and releasable multi-level caching mechanism for models.

1. Three-level caching: Disk, memory, and GPU memory form a logical cache queue. When switching models, the model is cached in the logical cache queue. This allows for faster model loading when the same model is used again. The memory and GPU memory caches use the LRU algorithm, while the disk cache uses the LFU algorithm. The number of cached models depends on the available memory and disk space specified. To enable the multi-level caching mechanism, use the `--arc` flag. To enable memory and disk caching, use `--system-ram-size` and `--arc-disk-size`, respectively.

2. GPU memory release: When generating large images, such as those with a resolution of 1024x1024, some models are cleared to free up GPU memory.

3. Handling CUDA out of memory: This multi-level caching mechanism is used stable diffusion in production situation. When a CUDA out of memory error occurs, the process is killed. This is to prevent cuda memory leaks caused by CUDA out of memory error, and with the help of an auto-restart script, the service availability can be maintained.

4. Setting and querying cached models: The `getCachedModels` API can be used to query the models cached in ram and GPU memory. The `setCachedModels` API can be used to set models that can be cached in memory or GPU memory.

Commit:
Merge branch 'arc' into 'dev1111'

multiple level cache

See merge request ai/sd-webui!9

Commit:
multiple level cache

Commit:
Merge branch 'dev' of https://ghproxy.com/github.com/AUTOMATIC1111/stable-diffusion-webui into dev1111

Commit:
code style fix

Commit:
code style fix

AUTOMATIC1111 commented:
PRs into this repo should be small and preferably do one thing.

There are multiple kinds of caching in the repo already, in CPU memory and in GPU memory.

I don't understand why you would ever have to write to disk for caching models.

The code explicitly refers to CUDA which likely means it will be incompatible for OSX users.

If there are GPU memory leaks, my vision is that they should be fixed rather than restarting the process.

Overall considering how this only injects itself into what seems to be just two places, `process_images_inner` + `reload_model_weights`, and is otherwise isolated from webui code, it seems perfectly suited for being an extension.

smallapes commented:
1. We have implemented a three-level caching mechanism for model objects not for checkpoint. Through real-world production testing, we have refined the functionality, so the code appears to be quite extensive. However, I am willing to delete some code, such as API queries, if they are deemed unnecessary for the plugin.

2. The aim of this caching solution is to optimize the time taken for model switching in production. The current code has two caching schemes. One caches model objects in either memory or cuda memory, with a fixed number of cache number. The other caches checkpoints in memory, also with a fixed number of number. These two schemes are isolated and may result in redundant cache objects. Moreover, when generating large images, cuda memory is not released automatically, leading to potential out-of-memory (OOM) issues. Additionally, caching checkpoints in memory can result in significant time spent on the "apply weights to model" step.

3. Hence, modifying the original caching scheme could potentially break existing functionality and is extremely challenging. Therefore, we have implemented a new approach with the following advantages: it is disabled by default, so it does not impact existing functionality when disabled, and the number of cache number is flexible based on available memory and cuda memory. The cache can also be cleared to generate large images.

4.If disk caching is not implemented, a new model loading may go through stages such as "create model", "load weights from disk", "apply weights to model", etc., which can be time-consuming, usually ranging from 3 to 8 seconds. If we cache the used models to disk using pickle, the loading time can be reduced to about 2 seconds. This is a significant optimization in production.

5. The three-level caching is disabled by default, so it does not impact OSX users. We have not yet tested it on the OSX system, as our main focus has been on Linux production servers.

6. The issue of GPU memory leaks already exists in the current code. For example, generating an image with dimensions of 2048x2048 and a batch size of 8 can cause this abnormality. Unfortunately, I haven't found a suitable method to handle this specific GPU memory leak, so the temporary solution in production is to kill the process and restart it.

7. Currently, we have only injected the modifications into the `reload_model_weights` function. If this approach is not considered elegant, I am more than willing to make changes. Alternatively, if you can provide more details on what the extension should specifically look like, I would greatly appreciate it.

AUTOMATIC1111 commented:
Extensions works like this: https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Developing-extensions

SirVeggie commented:
This sounds nice, has there been any progress on turning this into an extension?

AUTOMATIC1111 commented:
closing; reopen if there's more

AUTOMATIC1111 commented:
as usual a reminder: if some additional APIs are needed in main web UI to make the extension possible, they are welcome to be added in PRs
